packages/react-jss/src/withStyles.js

```diff
@@ -79,6 +77,61 @@ export default function withStyles<Theme: {}, S: Styles<Theme>>(
       // $FlowFixMe
       static defaultProps = {...InnerComponent.defaultProps}
 
+      static getSheetClasses(sheet, dynamicRules: ?DynamicRules) {
+        const classes = {}
+
+        // $FlowFixMe
+        for (const key in sheet.styles) {
+          classes[key] = sheet.classes[key]
+
+          if (dynamicRules && key in dynamicRules) {
+            classes[key] += ` ${sheet.classes[dynamicRules[key].key]}`
+          }
+        }
+
+        return classes
+      }
+
+      static updateDynamicRules(props: HOCProps<Theme, Props>, {dynamicRules, sheet}: State) {
+        if (!sheet) {
+          return
+        }
+
+        for (const key in dynamicRules) {
+          // $FlowFixMe: Not sure why it throws an error here
+          sheet.update(dynamicRules[key].key, props, {})
+        }
+      }
+
+      static removeDynamicRules(props: Props, {dynamicRules, sheet}: State) {
+        if (!sheet) {
+          return
+        }
+
+        for (const key in dynamicRules) {
+          sheet.deleteRule(dynamicRules[key].key)
+        }
+      }
+
+      static addDynamicStyles(sheet: StyleSheet): ?DynamicRules {
+        // $FlowFixMe Cannot access random fields on instance of class StyleSheet
+        if (!sheet.dynamicStyles) return undefined
```

Comment:
we could save dynamicStyles on a WeakMap, since we started using it anyways

Reply:
actually why not on state?

Reply:
Because we only need it once per sheet to be extracted

Reply:
right, then weakmap?

Reply:
Why would we need a WeakMap? There is only one dynamic styles object per sheet and we should save it on the sheet, we actually are already doing this.

Reply:
I did this because I didn't want to rely on WeakMap polyfill back then.  We could have a single WeakMap and use sheet as a key.

Reply:
Adding a property to a class instance is always a very dirty appoach. A less dirty would be to use a random property or a Symbol.

Reply:
But the sheet is already in a WeakMap. I don't think it makes sense to have another WeakMap.

Reply:
We are only using the sheet internally and the can't access it, so I don't think it's a big deal if we don't use a symbol.

Reply:
where?

Reply:
The SheetsManager is using a WeakMap now. Also, the dynamic styles shouldn't be GC collected as long as the sheet is still in memory.

Reply:
It won't be collected as long as the sheet is referenced.  I still think we should find a way to not attach properties to the sheet instance.

Reply:
Well yeah, it won’t be collected until the sheet is received and I think this is exactly what we want. Not attaching it would be nice, but it would also come with a downside. Mainly the extra time to find the dynamic styles in another WeakMap

Reply:
Are you talking about map.get(ref) perf cost? This is so fast we don't even need to talk about it.

Reply:
I think this would add an extra layer of complexity to it

Reply:
basically dynamicStyles = new WeakMap and dynamicStyles.get(sheet) or am I missing somethign?

Reply:
Yeah, we would need three WeakMaps because we also now need to track the static styles and the number of dynamic rules, so I don't think it would make sense to have these right now.